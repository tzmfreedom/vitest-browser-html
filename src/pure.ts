import { type FsOptions, type LocatorSelectors, server } from "@vitest/browser/context";
import { debug, getElementLocatorSelectors, type PrettyDOMOptions } from "@vitest/browser/utils";

const { readFile } = server.commands;

const mountedContainers = new Set<Element>();

export interface RenderResult extends LocatorSelectors {
  container: Element;
  debug(maxLength?: number, options?: PrettyDOMOptions): void;
}

export async function render(src: string): Promise<RenderResult> {
  if (src.trim().startsWith('<')) {
    return renderString(src);
  }
  return await renderFile(src)
}

export function renderString(html: string): RenderResult {
  const template = document.createElement('template');
  template.innerHTML = html;
  const container = template.content.firstElementChild;
  if (container === null) {
    throw new Error('failed rendering with document.createElement');
  }
  document.body.appendChild(container);
  mountedContainers.add(container);

  return {
    container,
    debug: (maxLength, options) => debug(container, maxLength, options),
    ...getElementLocatorSelectors(container),
  };
}

export async function renderFile(
  file: string,
  options?: BufferEncoding | FsOptions,
): Promise<RenderResult> {
  const body = await readFile(file, options);
  return renderString(body.toString());
}

export function cleanup() {
  for (const mountedContainer of mountedContainers) {
    if (mountedContainer.parentNode === document.body) {
      document.body.removeChild(mountedContainer);
    }
  }
}

