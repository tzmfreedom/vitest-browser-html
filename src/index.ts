import { beforeEach } from 'vitest';
import { page } from '@vitest/browser/context';
import { cleanup, render } from './pure';

export { render, renderString, renderFile, cleanup } from './pure';
export type { RenderResult } from './pure'

page.extend({
  render,
  [Symbol.for('vitest:component-cleanup')]: cleanup,
})

beforeEach(() => {
  cleanup();
});


declare module '@vitest/browser/context' {
  interface BrowserPage {
    render: typeof render
  }
}
