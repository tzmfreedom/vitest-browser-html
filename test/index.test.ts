import { expect, test } from 'vitest';
import { render, renderString, renderFile } from '../src';
import { page } from '@vitest/browser/context';
import "@vitest/browser/matchers";

test('renderString', async () => {
  const screen = renderString('<div>String</div>');
  await expect.element(screen.getByText('String')).toBeVisible();
  expect(screen.container).toMatchSnapshot();
});

test('renderFile', async () => {
  const screen = await renderFile('./fixtures/index.html');
  await expect.element(screen.getByText('HTMLFile')).toBeVisible();
  expect(screen.container).toMatchSnapshot();
});

test('render: string', async () => {
  const screen = await render('<div>String</div>');
  await expect.element(screen.getByText('String')).toBeVisible();
  expect(screen.container).toMatchSnapshot();
});

test('render: file', async () => {
  const screen = await render('./fixtures/index.html');
  await expect.element(screen.getByText('HTMLFile')).toBeVisible();
  expect(screen.container).toMatchSnapshot();
});

test('page.render: file', async () => {
  const screen = await page.render('./fixtures/index.html');
  await expect.element(screen.getByText('HTMLFile')).toBeVisible();
  expect(screen.container).toMatchSnapshot();
});