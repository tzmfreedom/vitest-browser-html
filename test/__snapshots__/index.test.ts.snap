// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`page.render: file 1`] = `
<div>
  HTMLFile
</div>
`;

exports[`render html file 1`] = `
<div>
  HTMLFile
</div>
`;

exports[`render html string 1`] = `
<div>
  String
</div>
`;

exports[`render: file 1`] = `
<div>
  HTMLFile
</div>
`;

exports[`render: string 1`] = `
<div>
  String
</div>
`;

exports[`renderFile 1`] = `
<div>
  HTMLFile
</div>
`;

exports[`renderString 1`] = `
<div>
  String
</div>
`;
